client/src/use-query.tsx

```diff
@@ -0,0 +1,26 @@
+import {Entity, generateZQL} from '@rocicorp/rails';
+import {EntityQuery} from '@rocicorp/rails/out/zql/query/entity-query.js';
+import {EntitySchema} from '@rocicorp/rails/out/zql/schema/entity-schema.js';
+import {useEffect, useMemo, useState} from 'react';
+
+export function useQuery<S extends EntitySchema, R>(
+  query: EntityQuery<S, R>,
+  dependencies: unknown[] = [],
+): R {
+  const view = useMemo(() => query.prepare().materialize(), dependencies);
+  const [value, setValue] = useState<R>(view.value as R);
+  useEffect(() => {
```

Comment:
Setting state via `useEffect` causes UI flashing / tearing.

`useEffect` will not run in the same render pass as the render caused by a change in dependencies. 

As in the component will render and fully paint to the browser once with the updated dependencies and then again with the affects (setValue) of `useEffect`.

You'd need to emulate `useEffect` with `useState` to prevent flashing.

```
const [prevDependencies, setPrevDependencies] = useState(dependencies);
if (!shallowEqual(dependencies, prevDependencies)) {
  setPrevDependencies(prevDependencies);
  const newView = viewFn();
  setValue(newView.value);
  setView(newView);
  disposer();
  setDisposer(newView.on(setValue));
}
```

This will ensure that when dependencies change, the new view is created and value set on the component before it renders to the screen.

Reply:
some related discussion: https://twitter.com/tantaman/status/1732140032729985512

Reply:
You have certainly opened up a can of worms!

The code that you have here and the code on twitter seems incomplete. We need to have a cleanup phase too to remove the listener.

Reply:
See also `useLayoutEffect` ?

Reply:
> The code that you have here and the code on twitter seems incomplete. We need to have a cleanup phase too to remove the listener.

 `setDisposer` would handle that, although given the `disposer` isn't used to render anything it should be a `ref` rather than state.

This should be more complete:

```ts
// track previous deps
const [prevDependencies, setPrevDependencies] = useState();
// track our view / query results
const [view, setView] = useState();
// track our subscription to the view
const [disposer, setDisposer] = useState();

// prevDeps === undefined -> first render
// prevDeps != deps, re-render
if (prevDependencies === undefined || !shallowEqual(dependencies, prevDependencies)) {
  setPrevDependencies(prevDependencies);
  // new deps? compute view / query
  const newView = viewFn();
  // destroy old view
  view?.destroy();
  // destroy old subscription
  disposer?.();

  setValue(newView.value);
  setView(newView);
  setDisposer(newView.on(setValue));
}

// also need to try cleaning up on un-mount of the component
useEffect(() => {
  disposer?.();
  view?.destroy();
}, []);
```

Reply:
Thanks. I will try again to see if I can get something along these lines working.

For replicache-react, useSubscribe was simpler because useSubscribe is always async and we do not have the result immediately.